midenc-compile/src/stages/cross_ctx/lower_imports.rs

```diff
@@ -0,0 +1,281 @@
+//! lowering the imports into the Miden ABI for the cross-context calls
+
+use std::collections::{BTreeMap, VecDeque};
+
+use miden_assembly::Spanned;
+use midenc_hir::{
+    diagnostics::Severity, pass::AnalysisManager, types::Abi, Block, Call, ComponentBuilder,
+    ComponentImport, Function, FunctionIdent, FunctionType, InstBuilder, Instruction,
+    MidenAbiImport, Signature, SourceSpan, Symbol, UnsafeRef,
+};
+use midenc_session::{DiagnosticsHandler, Session};
+
+use super::flat::{
+    assert_core_wasm_signature_equivalence, flatten_function_type, needs_transformation,
+    FlatteningDirection,
+};
+use crate::{stage::Stage, CompilerResult, LinkerInput};
+
+/// Generates lowering for imports for the cross-context calls according to the Miden ABI.
+///
+/// For each component import ensure a module for each interface that has a lowering function i.e. a
+/// function that takes the arguments according to the Wasm CABI, converts them to the cross-context
+/// Miden ABI, calls the imported function with the lowered arguments, takes the result
+/// (cross-context Miden ABI), and converts it to the Wasm CABI.
+/// The calls to the component import is switched to the generated lowering function.
+pub struct LowerImportsCrossCtxStage;
+
+impl Stage for LowerImportsCrossCtxStage {
+    type Input = LinkerInput;
+    type Output = LinkerInput;
+
+    fn run(
+        &mut self,
+        input: Self::Input,
+        analyses: &mut AnalysisManager,
+        session: &Session,
+    ) -> CompilerResult<Self::Output> {
+        let component = if let LinkerInput::Hir(component) = input {
+            component
+        } else {
+            return Ok(input);
+        };
+
+        // dbg!(&component.imports());
+        // dbg!(&component.modules().keys());
+
+        let mut component_builder = ComponentBuilder::load(*component, &session.diagnostics);
+
+        let mut lowered_imports: BTreeMap<FunctionIdent, ComponentImport> = BTreeMap::new();
+        let imports = component_builder.imports().clone();
+        for (core_import_func_id, import) in imports.into_iter() {
+            if let ComponentImport::MidenAbiImport(_) = import {
```

Comment:
*Note:* All imports in a core Wasm module (whether they are imports of a `canon lower` declaration or of another core module's exports) have the same ABI (in terms of type signature and semantics) from a caller point of view. IMO we do not need to distinguish between them. It should be noted that an import of another component's exports, can only ever be called if it is synthesized via `canon lower`. In effect, the three ABIs I mentioned in my other comment are sufficient to encode all that information, and we should only be synthesizing lowering functions for `canon lower` declarations, so we don't need to consider ABI here at all - though that may have to wait until HIR2 is being integrated.

Reply:
Yes. I intend to ditch the component import/exports types in their current form during the HIR2 integration. 
This one is used to make the stage idempotent(re-entrable) in case of the added imports later, plus to skip the remnants of the SDK/intrinsics (`heap_base`) which should not be in the component imports to begin with.